yjit/src/codegen.rs

```diff
@@ -5315,11 +5367,121 @@ fn gen_send_general(
                 match opt_type {
                     OPTIMIZED_METHOD_TYPE_SEND => {
-                        gen_counter_incr!(asm, send_optimized_method_send);
-                        return CantCompile;
+
+                        // This is for method calls like `foo.send(:bar)`
+                        // The `send` method does not get its own stack frame.
+                        // instead we look up the method and call it,
+                        // doing some stack shifting based on the VM_CALL_OPT_SEND flag
+
+                        let starting_context = *ctx;
+
+                        if argc == 0 {
+                            gen_counter_incr!(asm, send_send_wrong_args);
+                            return CantCompile;
+                        }
+
+                        argc -= 1;
+
+                        let compile_time_name = jit_peek_at_stack(jit, ctx, argc as isize);
+
+                        if !compile_time_name.string_p() && !compile_time_name.static_sym_p()  {
+                            gen_counter_incr!(asm, send_send_chain_not_string_or_sym);
+                            return CantCompile;
+                        }
+
+                        mid = unsafe { rb_get_symbol_id(compile_time_name) };
+                        if mid == 0 {
+                            gen_counter_incr!(asm, send_send_null_mid);
+                            return CantCompile;
+                        }
+
+                        cme = unsafe { rb_callable_method_entry(comptime_recv_klass, mid) };
+                        if cme.is_null() {
+                            gen_counter_incr!(asm, send_send_null_cme);
+                            return CantCompile;
+                        }
+
+                        // We aren't going to handle `send(send(:foo))`. We would need to
+                        // do some stack manipulation here or keep track of how many levels
+                        // deep we need to stack manipulate
+                        // Because of how exits currently work, we can't do stack manipulation
+                        // until we will no longer side exit.
+                        let def_type = unsafe { get_cme_def_type(cme) };
+                        if let VM_METHOD_TYPE_OPTIMIZED = def_type {
+                            let opt_type = unsafe { get_cme_def_body_optimized_type(cme) };
+                            if let OPTIMIZED_METHOD_TYPE_SEND = opt_type {
+                                gen_counter_incr!(asm, send_send_nested);
+                                return CantCompile;
+                            }
+                        }
+
+                        let kw_splat = if flags & VM_CALL_KW_SPLAT != 0 {
+                            VM_CALL_KW_SPLAT
+                        } else {
+                            0
+                        };
+                        flags = VM_CALL_FCALL | VM_CALL_OPT_SEND | kw_splat;
```

Comment:
I think all you want here is `flags |= VM_CALL_OPT_SEND` instead of discarding all the old flags. It's not really a problem currently because we sometimes don't check the passed in flags in subsequent calls but we should be doing that.

Reply:
Hmm, that's not how I read https://github.com/ruby/ruby/blob/master/vm_insnhelper.c#L3368 but I might be missing something.

Reply:
It's true that the flags are reset in the interpreter, but the way it works relates to how block args and keyword args are handled. For example for keyword args, by the time it gets to the point where it resets the flags, some processing is already done in CALLER_SETUP_ARG. In YJIT we don't do any processing until later so we can't really discard this information.

I said there is no problem currently but there is actually a behavior difference on your branch because of the flags handling. Try the following with and without `--yjit-call-threshold=1`:

```ruby
class A
  attr_accessor :foo
end

def use(a)
  a.send(:foo=, bar: 3)
  p a.foo
end

use A.new
```